Measured wires on qscript

comp-phys-marc opened the pull request:
### Before submitting

Please complete the following checklist when submitting a PR:

- [✓] All new features must include a unit test.
      If you've fixed a bug or added code that should be tested, add a test to the
      test directory!

- [✓] All new functions and code must be clearly commented and documented.
      If you do make documentation changes, make sure that the docs build and
      render correctly by running `make docs`.

- [✓] Ensure that the test suite passes, by running `make test`.

- [✓] Add a new entry to the `doc/releases/changelog-dev.md` file, summarizing the
      change, and including a link back to the PR.

- [✓] The PennyLane source code conforms to
      [PEP8 standards](https://www.python.org/dev/peps/pep-0008/).
      We check all of our code against [Pylint](https://www.pylint.org/).
      To lint modified files, simply `pip install pylint`, and then
      run `pylint pennylane/path/to/file.py`.

When all the above are checked, delete everything above the dashed
line and fill in the pull request template.

------------------------------------------------------------------------------------------------------------

**Context:** A nice-to-have feature of the `QuantumScript` requested in issue #3203.

**Description of the Change:**  Adds a cached property to `QuantumScript` for the measured wires.

**Benefits:** The wires can be referenced instead of re-calculated, for example in `_update_observables` and `QuantumScript().measured_wires`. 

**Possible Drawbacks:** The data footprint of a `QuantumScript` object is now larger.

**Related GitHub Issues:** https://github.com/PennyLaneAI/pennylane/issues/3203

albi3ro commented on pennylane/tape/qscript.py:
```diff
@@ -438,6 +449,25 @@ def _update_observables(self):
+                    self._measured_wires.append(wire)
+
+        for m in self.measurements:
+            if isinstance(m.obs, list):
```
I don't think `m.obs` would ever be a list.  It should only be either `None` or an `Operator`.

albi3ro commented on pennylane/tape/qscript.py:
```diff
@@ -438,6 +449,25 @@ def _update_observables(self):
+            _measured_wires (int): The wires measured in this QuantumScript
+        """
+
+        def extract_wires(obs):
+            for wire in obs.wires:
+                if wire not in self._measured_wires:
+                    self._measured_wires.append(wire)
```
Might be good to point you toward [`Wires.all_wires`](https://docs.pennylane.ai/en/stable/code/api/pennylane.wires.Wires.html#pennylane.wires.Wires.all_wires)

comp-phys-marc replied:
Thank you, I can indeed use `Wires.all_wires` here instead if you prefer.

albi3ro commented on tests/tape/test_qscript.py:
```diff
@@ -128,6 +130,21 @@ def test_cached_graph_specs_reset(self):
+        measurement = [
+            qml.probs(wires=(-1)),
+            qml.expval(qml.Hermitian(2 * np.eye(2), wires=-2)),
+            qml.PauliX(-1),
```
While this is unvalidated during the initialization process,  `measurements` should be [`Iterable[MeasurementProcess]`](https://github.com/PennyLaneAI/pennylane/blob/597beac35f49874a042f6e0dc61bdff69cce19a3/pennylane/tape/qscript.py#L8). We don't perform this validation since `QuantumScript` is mostly internal and produced in controlled situations.

`PauliX` is an `Operator`, not a `MeasurementProcess`, so it shouldn't be in the `measurement` list.

comp-phys-marc replied:
That's interesting. I am just getting to know Pennylane so thanks for bearing with me. Is it the case that I need to wrap it in qml.expval to get the expectation value of the observable?

albi3ro replied:
Exactly.  `qml.expval(qml.PauliX(-1))` indicates taking an expectation value measurement of the observable `PauliX`.

albi3ro reviewed (commented):
Thanks @comp-phys-marc 🎉 

Make sure you also add your name to the contributors section at the bottom of the changelog and add a link to the entry.  This kind of change should also fall under "Improvements" rather than "New features". New Features tend to be larger, user-focused things.


While this is how we currently tend to treat a lot of the properties in `QuantumScript`, this may be a good place where we can start using a more "lazy" approach to the properties. 

We currently create a lot of "intermediate" quantum scripts, that are then transformed into something else. So for the vast majority of quantum scripts created, we won't be ever using `measured_wires`, so computing it is wasted time.

So I'd like to consider only computing `measured_wires` when we actually are going to use it. It's fairly lightweight, so this probably isn't that big of a deal, but might as well anyway.

Some options for being more lazy:

1) Recalculate it every time we need it: See [`samples_computational_basis`](https://github.com/PennyLaneAI/pennylane/blob/597beac35f49874a042f6e0dc61bdff69cce19a3/pennylane/tape/qscript.py#L313)

2) Leave `_measured_wires` as `None` till we calculate it for the first time: See [`specs`](https://github.com/PennyLaneAI/pennylane/blob/597beac35f49874a042f6e0dc61bdff69cce19a3/pennylane/tape/qscript.py#L1131) or `graph`.

3) Use `functools.cached_property`.  I haven't tried this out yet myself. But its an option.

I'm leaning toward option 1 till we start actually using the `measured_wires` property more often.

comp-phys-marc commented:
> Thanks @comp-phys-marc tada
> 
> Make sure you also add your name to the contributors section at the bottom of the changelog and add a link to the entry. This kind of change should also fall under "Improvements" rather than "New features". New Features tend to be larger, user-focused things.
> 
> While this is how we currently tend to treat a lot of the properties in `QuantumScript`, this may be a good place where we can start using a more "lazy" approach to the properties.
> 
> We currently create a lot of "intermediate" quantum scripts, that are then transformed into something else. So for the vast majority of quantum scripts created, we won't be ever using `measured_wires`, so computing it is wasted time.
> 
> So I'd like to consider only computing `measured_wires` when we actually are going to use it. It's fairly lightweight, so this probably isn't that big of a deal, but might as well anyway.
> 
> Some options for being more lazy:
> 
>     1. Recalculate it every time we need it: See [`samples_computational_basis`](https://github.com/PennyLaneAI/pennylane/blob/597beac35f49874a042f6e0dc61bdff69cce19a3/pennylane/tape/qscript.py#L313)
> 
>     2. Leave `_measured_wires` as `None` till we calculate it for the first time: See [`specs`](https://github.com/PennyLaneAI/pennylane/blob/597beac35f49874a042f6e0dc61bdff69cce19a3/pennylane/tape/qscript.py#L1131) or `graph`.
> 
>     3. Use `functools.cached_property`.  I haven't tried this out yet myself. But its an option.
> 
> 
> I'm leaning toward option 1 till we start actually using the `measured_wires` property more often.

Thanks so much for your feedback! I've made the property "lazy-loaded". I've also moved the new line in the Changelog to the Improvements section, added my name and addressed the rest of your notes. :)

timmysilv commented on pennylane/tape/qscript.py:
```diff
@@ -438,6 +447,20 @@ def _update_observables(self):
+            if m.obs is not None:
+                wires.append(m.obs.wires)
+
+        self._measured_wires = Wires.all_wires([*wires, self._measured_wires])
```
why do you do a union with `self._measured_wires`? this gets messy because QuantumScript is mutable, so it might be created measuring wires `[0, 1]`, then later be updated to measure `[2]`. With this code, if someone calls `qscript.measured_wires`, they'll get `[0, 1, 2]` but should probably only get `[2]`.

The challenge here is really deciding when to re-compute this attribute, because we don't exactly have a listener checking if `self.measurements` has been changed.

comp-phys-marc replied:
Thanks for raising this. It seems to me to be a use case issue that would be resolved by a best practice. Is there documentation for using QuantumScript that specifies when to mutate it versus when to instantiate a new one?

I would probably say that the user is taking things into their own hands when they mutate a QuantumScript. The question pertaining to this ticket seems to be whether the measured wires property remembers all the wires that are measured by the QuantumScript over time or not. Of course when dealing with a cached property, this is a pertinent question. I am open to whatever you think is the best behaviour.

timmysilv replied:
haha you're touching on questions I'm scared to answer! some thoughts:
- I agree that the user is taking things into their own hands when mutating. On the one hand, QuantumScript is a rather internal detail (hence why it's in the "Internal" section in the docs sidebar 😄) so typical users shouldn't be near it. On the other hand, it happens anyway. So I would strive to make it as clear as possible to the user that they are in fact mutating an existing object, and they know what the consequences are. Once informed, PennyLane has done its job and they can do what they like!
- I don't think PL devs have ever thought about a QuantumScript as something that evolves with time, so I'm inclined to say that if a user _measured_ some wires, changed observables, then _measured_ again, they should probably make a new tape. Tapes often undergo various transformations, but I think usually before execution, and not in this way.

It's not totally clear to me what's best, so I will let you decide! Please make sure to document the assumption that you make, and I'm fine either way

comp-phys-marc replied:
Thanks for your thoughts. In this case I may just leave it as-is and assume as you have said that the QuantumScript won't evolve with time.

comp-phys-marc replied:
I've made this change in a recent commit.

timmysilv commented on pennylane/tape/qscript.py:
```diff
@@ -365,6 +373,7 @@ def _update(self):
         # The following line requires _par_info to be up to date
         self._update_trainable_params()  # Updates the _trainable_params; O(1)
 
+        self._update_measured_wires()
```
this method is admittedly misleading, but by default, `_update` is called on tape construction, so putting this here would make this evaluation eager (non-lazy).
```suggestion
```

comp-phys-marc replied:
I've accepted your suggested change.

timmysilv commented on pennylane/tape/qscript.py:
```diff
@@ -226,6 +228,12 @@ def __len__(self):
+    @property
+    def measured_wires(self) -> List[int]:
+        """Returns the measured wires"""
+        self._update_measured_wires()
```
as mentioned below, the question is to decide when we need to re-evaluate this. I'm fine with re-evaluating whenever someone requests it, but at that point there is no need for the private `_measured_wires` property - just return whatever you evaluate!

comp-phys-marc replied:
This was done based on [albi3ro](https://github.com/albi3ro)'s feedback above. He pointed to `samples_computational_basis` as an example for the desired behaviour.

timmysilv replied:
sorry, I think we've got our wires crossed 😆 the `@property` attribute is perfect, I'm more saying that you don't need to save the value to `_measured_wires`. Instead, you can do:
```python
@property
def measured_wires(self) -> Wires:
    """Returns the measured wires"""
    wires = []
    for m in self.measurements:
        if m.obs is not None:
            wires.append(m.obs.wires)
    return Wires.all_wires([*wires, self._measured_wires])
```
and delete the `_update_measured_wires` private method. See the `observables` property for something similar.

comp-phys-marc replied:
Now that `_update_measured_wires` isn't used in `_update`, this makes sense.

comp-phys-marc replied:
I've made this change in a recent commit.

timmysilv reviewed (commented):
thanks for putting this into motion, it's looking good!

when a decision is made re: when to evaluate, make sure to add tests that show the re-evaluation working (ie. does/doesn't happen as expected in certain cases).

comp-phys-marc commented:
Hello, I have updated and merged in main. I am wondering if there is anything else that needs be done for this PR to be reviewed? Thank you. :)

timmysilv commented on doc/releases/changelog-dev.md:
```diff
@@ -404,6 +404,13 @@
 * A small typo was fixed in the docstring for `qml.sample`.
   [(#5685)](https://github.com/PennyLaneAI/pennylane/pull/5685)
 
+* `qml.operation.WiresEnum.AllWires` is now -2 instead of 0 to avoid the
+  ambiguity between `op.num_wires = 0` and `op.num_wires = AllWires`.
+  [(#3978)](https://github.com/PennyLaneAI/pennylane/pull/3978)
+  
```
```suggestion
```

timmysilv commented on doc/releases/changelog-dev.md:
```diff
@@ -538,6 +545,7 @@
 
 This release contains contributions from (in alphabetical order):
 
+Komi Amiko,
```
```suggestion
```

timmysilv commented on doc/releases/changelog-dev.md:
```diff
@@ -563,6 +571,8 @@ Lee James O'Riordan,
 Mudit Pandey,
 Kenya Sakka,
 Jay Soni,
+David Wierichs,
```
```suggestion
```

timmysilv commented on pennylane/tape/qscript.py:
```diff
@@ -31,6 +31,8 @@
 from pennylane.typing import TensorLike
 from pennylane.wires import Wires
 
+_empty_wires = qml.wires.Wires([])
+
```
```suggestion
```

timmysilv commented on tests/tape/test_qscript.py:
```diff
@@ -456,6 +457,32 @@ def test_iteration_preserves_circuit(self):
+        return QuantumScript(prep + ops, measurement)
+
+    def test_measured_wires(self, make_script):
+        """Test that measured_wires property is set when called and not before."""
+        qs = make_script
+
+        assert len(qs._measured_wires) == 0
```
```suggestion
        """Test the measured_wires property."""
        qs = make_script
```

timmysilv reviewed (commented):
Hi again @comp-phys-marc! The lazy-loading looks great 😎 there are still some remnants of the property caching that we aren't really benefiting from here, and I don't think we need to do that anyway. Plus, if we feel it would add value in the future, decorating with `@functools.cached_property` instead of `@property` as @albi3ro suggested should work perfectly. I've left some inline comments to help tidy those things up. Let me know if you've got any questions!

trbromley commented:
Hi @comp-phys-marc! Thanks a lot for opening this PR and returning back to it after a few months.

We've had an internal discussion about the utility of having a measured wires attribute, as requested in the original #3203 issue. Since that issue was opened, things have evolved a bit. In fact, we're moving towards a new program representation - you can check it out [here](https://github.com/PennyLaneAI/pennylane/tree/96a3d56de81bf7cf2d7a98786edef829e5bdf53d/pennylane/capture). While [QuantumScript](https://docs.pennylane.ai/en/stable/code/api/pennylane.tape.QuantumScript.html) will still exist for the foreseeable future, we're cautious about changing it up during this transition period because we'd like to avoid the entropy and additional maintenance burden.

Because of this, we think that it would be best to close #3203 and this PR. Sorry for this, we know that it might be a bit frustrating given the original request in #3203. However, we still love to work with contributors like yourself and could suggest a fresher issue to work on if you would like.

comp-phys-marc commented:
Hello, thanks so much for getting back to me on this! It's no problem, and it's exciting that you are moving to a new program representation. Please do feel free to suggest a fresher issue for me to work on. :)
